### PR TITLE
DSP-1161 Add error handling in the new resource form

### PR DESCRIPTION
**Pull request by flavens**
resolves DSP-1161 + DSP-852

### CODE REVIEW

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts [hunk 1]
```diff
@@ -184,6 +187,11 @@ export class ResourceInstanceFormComponent implements OnInit, OnDestroy {
                     for (const project of response.body.user.projects) {
                         this.usersProjects.push(project);
                     }
+
+                    // notifies the user that he/she is not part of any project
+                    if (this.usersProjects.length === 0) {
+                        this.errorMessage = 'You are not part of any activated project.';
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts:193]: 'You are not a part of any active projects.'

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts [hunk 2]
```diff
@@ -234,14 +245,19 @@ export class ResourceInstanceFormComponent implements OnInit, OnDestroy {
                         response.ontologies = response.ontologies.filter(onto => onto.attachedToProject !== Constants.SystemProjectIRI);
 
                         this.ontologiesMetadata = response;
+
+                        // notifies the user that the select project has got 0 data model yet defined.
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts:249]: notifies the user that the selected project does not have any data models defined yet.

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts [hunk 3]
```diff
@@ -234,14 +245,19 @@ export class ResourceInstanceFormComponent implements OnInit, OnDestroy {
...
+                        if (!this.selectOntologyComponent && response.ontologies.length === 0) {
+                            this.errorMessage = 'No data model defined for the select project.';
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts:251]: models

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts [hunk 4]
```diff
@@ -234,14 +245,19 @@ export class ResourceInstanceFormComponent implements OnInit, OnDestroy {
...
+                        }
                     },
                     (error: ApiResponseError) => {
                         console.error(error);
                     }
                 );
             }
         } else {
-            this.errorMessage = 'You are not part of any project.';
+            this.errorMessage = 'You are not part of any activated project.';
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts:260]: 'You are not a part of any active projects.'

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts [hunk 5]
```diff
@@ -285,6 +304,11 @@ export class ResourceInstanceFormComponent implements OnInit, OnDestroy {
                                 // since the component already exists, the ngAfterInit method of the component will not be called so we must assign the value here manually
                                 this.selectResourceClassComponent.form.controls.resources.setValue(this.resourceClasses[0].id);
                             }
+
+                            // notifies the user that the select ontology has got 0 resource class yet defined.
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts:308]: notifies the user that the selected ontology does not have any resource classes defined yet.

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts [hunk 6]
```diff
@@ -285,6 +304,11 @@ export class ResourceInstanceFormComponent implements OnInit, OnDestroy {
...
+                            if ((!this.selectResourceClassComponent || this.selectOntologyComponent.form.controls.ontologies.valueChanges) && this.resourceClasses.length === 0) {
+                                this.errorMessage = 'No resource defined for the select ontology.';
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts:310]: 'No resources defined for the selected ontology.'

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts [hunk 7]
```diff
@@ -325,6 +349,14 @@ export class ResourceInstanceFormComponent implements OnInit, OnDestroy {
 
                     // filter out all props that cannot be edited or are link props
                     this.properties = onto.getPropertyDefinitionsByType(ResourcePropertyDefinition).filter(prop => prop.isEditable && !prop.isLinkProperty);
+
+                    // notifies the user that the select resource has got 0 property yet defined.
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts:353]: notifies the user that the selected resource does not have any properties defined yet.

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts [hunk 8]
```diff
@@ -325,6 +349,14 @@ export class ResourceInstanceFormComponent implements OnInit, OnDestroy {
...
+                    if (!this.selectPropertiesComponent && this.properties.length === 0) {
+                        this.errorMessage = 'No property defined for the select resource.';
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts:355]: 'No properties defined for the selected resource.'

File: src/app/workspace/resource/resource-instance-form/select-ontology/select-ontology.component.html [hunk 9]
```diff
@@ -1,12 +1,18 @@
 <span *ngIf="ontologiesMetadata !== undefined" [formGroup]="form">
     <mat-form-field *ngIf="ontologiesMetadata.ontologies.length === 1" class="large-field selected-ontology">
-        <mat-label>Select an ontology</mat-label>
+        <mat-label>Select a data model *</mat-label>
         <input matInput disabled value="{{ontologiesMetadata.ontologies[0].label}}" class="onto-label">
+        <mat-error *ngIf="form.controls.ontologies.hasError('required')">
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/select-ontology/select-ontology.component.html:5]: If this is automatically chosen for the user, do we ever encounter a scenario where this error message is needed?

File: src/app/workspace/resource/resource-instance-form/select-project/select-project.component.html [hunk 10]
```diff
@@ -1,14 +1,20 @@
 <span *ngIf="form" [formGroup]="form">
     <mat-form-field *ngIf="usersProjects.length === 1" class="large-field selected-project">
-        <mat-label>Select a project</mat-label>
+        <mat-label>Select a project *</mat-label>
         <input matInput disabled value="{{usersProjects[0].shortcode}} | {{usersProjects[0].shortname}}" class="project-label">
+        <mat-error *ngIf="form.controls.projects.hasError('required')">
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/select-project/select-project.component.html:5]: same as above

File: src/app/workspace/resource/resource-instance-form/select-resource-class/select-resource-class.component.html [hunk 11]
```diff
@@ -1,16 +1,25 @@
 <span *ngIf="resourceClassDefinitions?.length > 0" [formGroup]="form">
     <mat-form-field *ngIf="resourceClassDefinitions.length === 1" class="large-field selected-res-class">
-        <mat-label>Select a resource</mat-label>
+        <mat-label>Select a resource *</mat-label>
         <input matInput disabled value="{{resourceClassDefinitions[0].label}}" class="res-label">
+        <mat-error *ngIf="form.controls.resources.hasError('required')">
```
Comment (mdelez) [inline src/app/workspace/resource/resource-instance-form/select-resource-class/select-resource-class.component.html:5]: same as above

**Review by mdelez — CHANGES_REQUESTED** [review]
I've found a scenario in which the 'Next' button is enabled when it shouldn't be. This might cause us a headache again 😰 

Screen Shot 2020-12-15 at 14 50 54

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts (hunk 1 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts]: fixed in 3717fc5

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts (hunk 3 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts]: fixed in 3717fc5

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts (hunk 2 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts]: fixed in 3717fc5

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts (hunk 4 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts]: fixed in 3717fc5

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts (hunk 6 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts]: fixed in 3717fc5

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts (hunk 7 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts]: fixed in 3717fc5

File: src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts (hunk 8 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/resource-instance-form.component.ts]: fixed in 3717fc5

File: src/app/workspace/resource/resource-instance-form/select-ontology/select-ontology.component.html (hunk 9 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/select-ontology/select-ontology.component.html:5]: fixed in 3717fc5

Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/select-ontology/select-ontology.component.html:5]: no you are right, I do not think that we will ever encounter such scenario

Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/select-ontology/select-ontology.component.html]: fixed in ec83af5

File: src/app/workspace/resource/resource-instance-form/select-project/select-project.component.html (hunk 10 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/select-project/select-project.component.html]: fixed in ec83af5

File: src/app/workspace/resource/resource-instance-form/select-resource-class/select-resource-class.component.html (hunk 11 above)
Comment (flavens) [reply src/app/workspace/resource/resource-instance-form/select-resource-class/select-resource-class.component.html]: fixed in ec83af5

**Comment by flavens** [comment]
> I've found a scenario in which the 'Next' button is enabled when it shouldn't be. This might cause us a headache again 😰

@mdelez would it make sense to disable the resource label field in this case? I do not see the point for the user to continue to fill in the form if there are no properties. If so, the form will stay invalid and the next button will stay disabled... not the best solution ever I agree

**Comment by mdelez** [comment]
> > I've found a scenario in which the 'Next' button is enabled when it shouldn't be. This might cause us a headache again 😰
> 
> @mdelez would it make sense to disable the resource label field in this case? I do not see the point for the user to continue to fill in the form if there are no properties. If so, the form will stay invalid and the next button will stay disabled... not the best solution ever I agree

That seems like a good solution to me 👍

**Review by mdelez — APPROVED** [review]
Looks good!

**Pull request merged**